### DISCUSSION TITLE
Get data of HTTP request in hooks

### DISCUSSION THREAD

**Discussion by CovertCode** [pocketbase/pocketbase#4077] [Q&A] [answered]
So when I update a record using API and a hook is triggered `$app.onRecordBeforeUpdateRequest("test").add((e)` how do I access the JSON data which was being sent to pocketbase server which triggered this hook?

**Comment by ganigeorgiev** [comment]
You shouldn't use `$app.onRecordBeforeUpdateRequest` as it is not concurrent safe.

Instead use the global hooks method shown in https://pocketbase.io/docs/js-event-hooks/#record-crud-api-hooks.

The request data can be accessed via the `e.httpContext`. For examples you can check https://pocketbase.io/docs/js-routing/#reading-request-body (_you can replace `c` from the examples with `e.httpContext`_).

**Reply by ganigeorgiev** [reply]
btw. If you found about `$app.onRecordBeforeUpdateRequest("test").add` from ChatGPT/Copilot/etc., please avoid blindly copying code from there as most likely it will be incorrect and/or trained on legacy data (_this is also why I don't encourage users working on tutorials as most likely they will be irrelevant once PocketBase v1 is released_).

**Reply by CovertCode** [reply]
I found it in pocketbase docs https://pocketbase.io/docs/js-records/#update-record-with-data-validations

I want to prevent field update even by admin as long as hook is enabled.
Was thinking something like this:
```
onRecordBeforeUpdateRequest((c) => {
     const order_name = c.formValue("order_name")
	 if(order_name != null || order_name != undefined || order_name != "") {
	 return null
	 } 
},"test")
```

or is it possible to remove `c.formValue("order_name")` from request so that this field is not updated by anyone?

**Reply by ganigeorgiev** [reply, marked as answer]
> I found it in pocketbase docs

Oh, I'll update the docs, thanks.

Couple things regarding your code snippet:
1. `c` in the above is a hook event and not the request context (I recommend renaming it to `e`). The request context can be accessed via `e.httpContext`.
2. Note that `e.httpContext.formValue` will work only with `multipart/form-data` requests, aka. it will not work if users submit `application/json`.
For unified access to the request body you can use `const data = $apis.requestInfo(e.httpContext).data` (data should be an object with the parsed request body).

Depending on your use case, a better alternative to check whether a field value has changed or not is to compare the new record value with the old one and throw an error:

```js
onRecordBeforeUpdateRequest((e) => {
    const original = e.record.originalCopy();

    const oldValue = original.get("yourField");
    const newValue = e.record.get("yourField");

    if (oldValue != newValue) {
        throw new BadRequestError("Failed to perform the request.", {
            yourField:  new ValidationError("immutable_error", "Cannot change the field value."),
        })
    }
}, "test")
```

**Reply by ganigeorgiev** [reply]
> or is it possible to remove c.formValue("order_name") from request so that this field is not updated by anyone?

No, because the request data is already populated in the record that is going to be saved at the end of the execution chain. If you want to silently ignore new values being set, you can modify the above example with:

```js
// always overwrite the record field value with its original one
e.record.set("yourField", original.get("yourField"));
```

**Reply by CovertCode** [reply]
Thanks. 
This seems to be optimal solution to prevent someone from updating a value.

Do you think it would be possible to create this hook via UI?
While adding a field to collection then there can be a button "prevent everyone from update"
Once this is selected then a hook file will be created as "collectionName_fieldName.pb.js"
With this field enabled or disabled this file will be removed from hooks folder.

This can be a feature in later stage like after v1 and I might be able to implement this in future and submit a PR to remove the load of implementing this.

**Reply by ganigeorgiev** [reply]
> Do you think it would be possible to create this hook via UI?

I'm not sure as the behavior for what a user consider "immutable" may differ from one use case to another (eg. a user may want to allow overwriting empty values or to allow admins performing any or limited actions on the field, etc.). It is hard to generalize in my opinion and for now this remains out of the scope.
